pkg/health/checker/checker.go

```diff
@@ -361,224 +180,347 @@ func watchConsulHealth(
 	}
 }
 
-func (h healthChecker) WatchService(
-	ctx context.Context,
+func (c consulHealthChecker) Service(
 	serviceID string,
-	resultCh chan<- map[types.NodeName]health.Result,
-	errCh chan<- error,
-	watchDelay time.Duration,
-) {
-	defer close(resultCh)
-	watchConsulHealth(ctx, serviceID, h.kv, resultCh, errCh, watchDelay)
+	_ manifest.StatusStanza,
+) (map[types.NodeName]health.Result, error) {
+	// return map[nodenames (string)] to consul.WatchResult
+	// get health of all instances of a service with 1 query
+	kvEntries, err := c.consulStore.GetServiceHealth(serviceID)
+	if err != nil {
+		return nil, err
+	}
+	ret := make(map[types.NodeName]health.Result)
+	for _, kvEntry := range kvEntries {
+		ret[kvEntry.Node] = consulWatchToResult(kvEntry)
+	}
+
+	return ret, nil
 }
 
-func statusURLToNodeName(s string) (types.NodeName, error) {
-	u, err := url.Parse(s)
-	if err != nil {
-		return "", util.Errorf("error parsing url '%s'", s)
+func consulWatchToResult(w consul.WatchResult) health.Result {
+	return health.Result{
+		ID:      w.Id,
+		Node:    w.Node,
+		Service: w.Service,
+		Status:  health.ToHealthState(w.Status),
 	}
-	host, _, err := net.SplitHostPort(u.Host)
+}
+
+func kvpToResult(kv api.KVPair) (*health.Result, error) {
+	res := &health.Result{}
+	err := json.Unmarshal(kv.Value, &res)
 	if err != nil {
-		return "", util.Errorf("error parsing host:port '%s'", u.Host)
+		return nil, util.Errorf("Could not unmarshal health at %s: %v", kv.Key, err)
 	}
-	return types.NodeName(host), nil
+	return res, nil
 }
 
-func healthResultsCopy(healthResults map[types.NodeName]health.Result) map[types.NodeName]health.Result {
-	out := make(map[types.NodeName]health.Result)
-	for nodeID, healthResult := range healthResults {
-		out[nodeID] = healthResult
+// Maps a list of KV Pairs into a slice of health.Results
+// Halts and returns upon encountering an error
+func kvpsToResult(kvs api.KVPairs) ([]*health.Result, error) {
+	result := make([]*health.Result, len(kvs))
+	for i, kv := range kvs {
+		res, err := kvpToResult(*kv)
+		if err != nil {
+			return nil, err
+		}
+		result[i] = res
 	}
-	return out
+
+	return result, nil
 }
 
-func (h shadowTrafficHealthChecker) WatchService(
+type HealthClient interface {
+	HealthCheckEndpoints(ctx context.Context, req *hclient.HealthEndpointsRequest) (map[string]health.HealthState, error)
+	HealthMonitor(ctx context.Context, req *hclient.HealthRequest, respCh chan *hclient.HealthResponse) error
+}
+
+type healthChecker struct {
```

Comment:
These would be easier to find if they were at the top of the file with the other types